File: app/src/main/java/me/devsaki/hentoid/database/CollectionDAO.java
package me.devsaki.hentoid.database;

import android.util.SparseIntArray;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.paging.PagedList;

import com.annimon.stream.function.Consumer;

import org.apache.commons.lang3.tuple.ImmutablePair;

import java.util.List;
import java.util.Map;
import java.util.Set;

import javax.annotation.Nullable;

import io.reactivex.Single;
import me.devsaki.hentoid.database.domains.Attribute;
import me.devsaki.hentoid.database.domains.Chapter;
import me.devsaki.hentoid.database.domains.Content;
import me.devsaki.hentoid.database.domains.ErrorRecord;
import me.devsaki.hentoid.database.domains.Group;
import me.devsaki.hentoid.database.domains.GroupItem;
import me.devsaki.hentoid.database.domains.ImageFile;
import me.devsaki.hentoid.database.domains.QueueRecord;
import me.devsaki.hentoid.database.domains.RenamingRule;
import me.devsaki.hentoid.database.domains.SearchRecord;
import me.devsaki.hentoid.database.domains.SiteBookmark;
import me.devsaki.hentoid.database.domains.SiteHistory;
import me.devsaki.hentoid.enums.AttributeType;
import me.devsaki.hentoid.enums.Grouping;
import me.devsaki.hentoid.enums.Site;
import me.devsaki.hentoid.enums.StatusContent;
import me.devsaki.hentoid.util.ContentHelper;
import me.devsaki.hentoid.util.SearchHelper;
import me.devsaki.hentoid.widget.ContentSearchManager;

public interface CollectionDAO {

    // CONTENT

    // Low-level operations

    LiveData<PagedList<Content>> selectNoContent();

    @Nullable
    Content selectContent(long id);

    List<Content> selectContent(long[] id);

    @Nullable
    Content selectContentByStorageUri(@NonNull final String folderUri, boolean onlyFlagged);

    @Nullable
    Content selectContentBySourceAndUrl(@NonNull Site site, @NonNull String contentUrl, @Nullable String coverUrl);

    Set<String> selectAllSourceUrls(@NonNull Site site);

    Set<String> selectAllMergedUrls(@NonNull Site site);

    List<Content> searchTitlesWith(@NonNull final String word, int[] contentStatusCodes);

    long insertContent(@NonNull final Content content);

    long insertContentCore(@NonNull final Content content);

    void updateContentStatus(@NonNull final StatusContent updateFrom, @NonNull final StatusContent updateTo);

    void updateContentDeleteFlag(long contentId, boolean flag);

    void deleteContent(@NonNull final Content content);

    List<ErrorRecord> selectErrorRecordByContentId(long contentId);

    void insertErrorRecord(@NonNull final ErrorRecord record);

    void deleteErrorRecords(long contentId);

    void clearDownloadParams(long contentId);

    void shuffleContent();


    // MASS OPERATIONS

    // Internal library (i.e. managed in the Hentoid folder)

    long countAllInternalBooks(boolean favsOnly);

    void streamAllInternalBooks(boolean favsOnly, Consumer<Content> consumer);

    void flagAllInternalBooks(boolean includePlaceholders);

    void deleteAllInternalBooks(boolean resetRemainingImagesStatus);

    // Queued books

    void flagAllErrorBooksWithJson();

    long countAllQueueBooks();

    LiveData<Integer> countAllQueueBooksLive();

    void deleteAllQueuedBooks();

    // Flagging

    void deleteAllFlaggedBooks(boolean resetRemainingImagesStatus);

    // External library

    long countAllExternalBooks();

    void deleteAllExternalBooks();


    // GROUPS

    List<Group> selectGroups(long[] groupIds);

    LiveData<List<Group>> selectGroupsLive(int grouping, @Nullable String query, int orderField, boolean orderDesc, int artistGroupVisibility, boolean groupFavouritesOnly, int filterRating);

    List<Group> selectGroups(int grouping, int subType);

    List<Group> selectGroups(int grouping);

    @Nullable
    Group selectGroup(long groupId);

    @Nullable
    Group selectGroupByName(int grouping, @NonNull final String name);

    long countGroupsFor(Grouping grouping);

    long insertGroup(Group group);

    void deleteGroup(long groupId);

    void deleteAllGroups(Grouping grouping);

    void flagAllGroups(Grouping grouping);

    void deleteAllFlaggedGroups();

    long insertGroupItem(GroupItem item);

    List<GroupItem> selectGroupItems(long contentId, Grouping grouping);

    void deleteGroupItems(List<Long> groupItemIds);


    // High-level queries (internal and external locations)

    List<Long> selectStoredContentIds(boolean nonFavouritesOnly, boolean includeQueued, int orderField, boolean orderDesc);

    long countStoredContent(boolean nonFavouriteOnly, boolean includeQueued);

    List<Content> selectContentWithUnhashedCovers();

    long countContentWithUnhashedCovers();


    void streamStoredContent(boolean nonFavouritesOnly, boolean includeQueued, int orderField, boolean orderDesc, Consumer<Content> consumer);


    Single<List<Long>> selectRecentBookIds(ContentSearchManager.ContentSearchBundle searchBundle);

    Single<List<Long>> searchBookIds(ContentSearchManager.ContentSearchBundle searchBundle, List<Attribute> metadata);

    Single<List<Long>> searchBookIdsUniversal(ContentSearchManager.ContentSearchBundle searchBundle);


    LiveData<PagedList<Content>> selectRecentBooks(ContentSearchManager.ContentSearchBundle searchBundle);

    LiveData<PagedList<Content>> searchBooks(ContentSearchManager.ContentSearchBundle searchBundle, List<Attribute> metadata);

    LiveData<PagedList<Content>> searchBooksUniversal(ContentSearchManager.ContentSearchBundle searchBundle);


    LiveData<List<Content>> selectErrorContentLive();

    LiveData<List<Content>> selectErrorContentLive(String query);

    List<Content> selectErrorContent();


    LiveData<Integer> countBooks(
            long groupId,
            List<Attribute> metadata,
            @ContentHelper.Location int location,
            @ContentHelper.Type int contentType);

    LiveData<Integer> countAllBooksLive();


    // IMAGEFILES

    void insertImageFile(@NonNull ImageFile img);

    void insertImageFiles(@NonNull List<ImageFile> imgs);

    void replaceImageList(long contentId, @NonNull final List<ImageFile> newList);

    void updateImageContentStatus(long contentId, StatusContent updateFrom, @NonNull StatusContent updateTo);

    void updateImageFileStatusParamsMimeTypeUriSize(@NonNull ImageFile image);

    void deleteImageFiles(@NonNull List<ImageFile> imgs);

    ImageFile selectImageFile(long id);

    List<ImageFile> selectImageFiles(long[] ids);

    LiveData<List<ImageFile>> selectDownloadedImagesFromContentLive(long id);

    List<ImageFile> selectDownloadedImagesFromContent(long id);

    Map<StatusContent, ImmutablePair<Integer, Long>> countProcessedImagesById(long contentId);

    Map<Site, ImmutablePair<Integer, Long>> selectPrimaryMemoryUsagePerSource();

    Map<Site, ImmutablePair<Integer, Long>> selectExternalMemoryUsagePerSource();


    // QUEUE

    List<QueueRecord> selectQueue();

    LiveData<List<QueueRecord>> selectQueueLive();

    LiveData<List<QueueRecord>> selectQueueLive(String query);

    void addContentToQueue(@NonNull final Content content, StatusContent targetImageStatus, @ContentHelper.QueuePosition int position, long replacedContentId, boolean isQueueActive);

    void updateQueue(@NonNull List<QueueRecord> queue);

    void deleteQueueRecordsCore();

    void deleteQueue(@NonNull Content content);

    void deleteQueue(int index);


    // ATTRIBUTES

    long insertAttribute(@NonNull Attribute attr);

    @Nullable
    Attribute selectAttribute(long id);

    Single<SearchHelper.AttributeQueryResult> selectAttributeMasterDataPaged(
            @NonNull List<AttributeType> types,
            String filter,
            long groupId,
            List<Attribute> attrs,
            @ContentHelper.Location int location,
            @ContentHelper.Type int contentType,
            boolean includeFreeAttrs,
            int page,
            int booksPerPage,
            int orderStyle);

    Single<SparseIntArray> countAttributesPerType(
            long groupId,
            List<Attribute> filter,
            @ContentHelper.Location int location,
            @ContentHelper.Type int contentType);


    // CHAPTERS

    List<Chapter> selectChapters(long contentId);

    void insertChapters(@NonNull final List<Chapter> chapters);

    void deleteChapters(@NonNull final Content content);

    void deleteChapter(@NonNull final Chapter chapter);


    // SITE HISTORY

    SiteHistory selectHistory(@NonNull Site s);

    void insertSiteHistory(@NonNull Site site, @NonNull String url);


    // BOOKMARKS

    long countAllBookmarks();

    List<SiteBookmark> selectAllBookmarks();

    List<SiteBookmark> selectBookmarks(@NonNull Site s);

    @Nullable
    SiteBookmark selectHomepage(@NonNull Site s);

    long insertBookmark(@NonNull SiteBookmark bookmark);

    void insertBookmarks(@NonNull List<SiteBookmark> bookmarks);

    void deleteBookmark(long bookmarkId);

    void deleteAllBookmarks();


    // SEARCH HISTORY

    LiveData<List<SearchRecord>> selectSearchRecordsLive();

    void insertSearchRecord(@NonNull SearchRecord record, int limit);

    void deleteAllSearchRecords();


    // RENAMING RULES

    @Nullable
    RenamingRule selectRenamingRule(long id);

    LiveData<List<RenamingRule>> selectRenamingRulesLive(@NonNull AttributeType type, String nameFilter);

    List<RenamingRule> selectRenamingRules(@NonNull AttributeType type, String nameFilter);

    long insertRenamingRule(@NonNull RenamingRule rule);

    void insertRenamingRules(@NonNull List<RenamingRule> rules);

    void deleteRenamingRules(List<Long> ids);

    void deleteAllRenamingRules();


    // RESOURCES

    void cleanup();

    long getDbSizeBytes();


    // ONE-TIME USE QUERIES (MIGRATION & CLEANUP)

    Single<List<Long>> selectOldStoredBookIds();

    long countOldStoredContent();

    long[] selectContentIdsWithUpdatableJson();
}

File: app/src/main/java/me/devsaki/hentoid/database/CollectionDAO.kt
package me.devsaki.hentoid.database

import android.util.SparseIntArray
import androidx.lifecycle.LiveData
import androidx.paging.PagedList
import me.devsaki.hentoid.core.Consumer
import me.devsaki.hentoid.database.domains.Attribute
import me.devsaki.hentoid.database.domains.Chapter
import me.devsaki.hentoid.database.domains.Content
import me.devsaki.hentoid.database.domains.ErrorRecord
import me.devsaki.hentoid.database.domains.Group
import me.devsaki.hentoid.database.domains.GroupItem
import me.devsaki.hentoid.database.domains.ImageFile
import me.devsaki.hentoid.database.domains.QueueRecord
import me.devsaki.hentoid.database.domains.RenamingRule
import me.devsaki.hentoid.database.domains.SearchRecord
import me.devsaki.hentoid.database.domains.SiteBookmark
import me.devsaki.hentoid.database.domains.SiteHistory
import me.devsaki.hentoid.enums.AttributeType
import me.devsaki.hentoid.enums.Grouping
import me.devsaki.hentoid.enums.Site
import me.devsaki.hentoid.enums.StatusContent
import me.devsaki.hentoid.util.AttributeQueryResult
import me.devsaki.hentoid.util.Location
import me.devsaki.hentoid.util.QueuePosition
import me.devsaki.hentoid.util.Type
import me.devsaki.hentoid.widget.ContentSearchManager.ContentSearchBundle

interface CollectionDAO {

    // CONTENT
    // Low-level operations
    fun selectNoContent(): LiveData<PagedList<Content>>

    fun selectContent(id: Long): Content?

    fun selectContent(id: LongArray): List<Content>

    fun selectContentByStorageUri(folderUri: String, onlyFlagged: Boolean): Content?

    fun selectContentByStorageRootUri(rootUri: String): List<Content>

    fun selectContentByUrlOrCover(site: Site, contentUrl: String, coverUrl: String?): Content?

    fun selectContentsByUrl(site: Site, contentUrl: String): Set<Content>

    fun selectContentsByQtyPageAndSize(qtyPage: Int, size: Long): Set<Content>

    fun selectAllSourceUrls(site: Site): Set<String>

    fun selectAllMergedUrls(site: Site): Set<String>

    fun searchTitlesWith(word: String, contentStatusCodes: IntArray): List<Content>

    fun insertContent(content: Content): Long

    fun insertContentCore(content: Content): Long

    fun updateContentStatus(updateFrom: StatusContent, updateTo: StatusContent)

    fun updateContentProcessedFlag(contentId: Long, flag: Boolean)

    fun updateContentsProcessedFlagById(contentIds: List<Long>, flag: Boolean)

    fun updateContentsProcessedFlag(contents: List<Content>, flag: Boolean)

    fun deleteContent(content: Content)

    fun selectErrorRecordByContentId(contentId: Long): List<ErrorRecord>

    fun insertErrorRecord(record: ErrorRecord)

    fun deleteErrorRecords(contentId: Long)

    fun clearDownloadParams(contentId: Long)

    fun shuffleContent()


    // MASS OPERATIONS
    // Primary library ("internal books")
    fun countAllInternalBooks(rootPath: String, favsOnly: Boolean): Long

    fun streamAllInternalBooks(rootPath: String, favsOnly: Boolean, consumer: Consumer<Content>)

    fun flagAllInternalBooks(rootPath: String, includePlaceholders: Boolean)

    fun deleteAllInternalContents(rootPath: String, resetRemainingImagesStatus: Boolean)

    // Queued books
    fun flagAllErrorBooksWithJson()

    fun countAllQueueBooks(): Long

    fun countAllQueueBooksLive(): LiveData<Int>

    fun deleteAllQueuedBooks()


    // Flagging
    fun deleteAllFlaggedContents(resetRemainingImagesStatus: Boolean, pathRoot: String?)


    // External library
    fun deleteAllExternalBooks()

    fun flagAllExternalContents()


    // GROUPS
    fun selectGroups(groupIds: LongArray): List<Group>

    fun selectGroupsLive(
        grouping: Int,
        query: String?,
        orderField: Int,
        orderDesc: Boolean,
        artistGroupVisibility: Int,
        groupFavouritesOnly: Boolean,
        groupNonFavouritesOnly: Boolean,
        filterRating: Int
    ): LiveData<List<Group>>

    fun selectGroups(grouping: Int): List<Group>

    fun selectGroups(grouping: Int, subType: Int): List<Group>

    fun selectEditedGroups(grouping: Int): List<Group>

    fun selectGroup(groupId: Long): Group?

    fun selectGroupByName(grouping: Int, name: String): Group?

    fun countGroupsFor(grouping: Grouping): Long

    fun insertGroup(group: Group): Long

    fun deleteGroup(groupId: Long)

    fun deleteAllGroups(grouping: Grouping)

    fun flagAllGroups(grouping: Grouping)

    fun deleteAllFlaggedGroups()

    fun deleteEmptyArtistGroups()

    fun insertGroupItem(item: GroupItem): Long

    fun selectGroupItems(contentId: Long, grouping: Grouping): List<GroupItem>

    fun deleteGroupItems(groupItemIds: List<Long>)


    // High-level queries (internal and external locations)
    fun selectStoredFavContentIds(bookFavs: Boolean, groupFavs: Boolean): Set<Long>

    fun selectContentWithUnhashedCovers(): List<Content>

    fun countContentWithUnhashedCovers(): Long

    fun streamStoredContent(
        includeQueued: Boolean,
        orderField: Int,
        orderDesc: Boolean,
        consumer: Consumer<Content>
    )


    fun selectRecentBookIds(searchBundle: ContentSearchBundle): List<Long>

    fun searchBookIds(searchBundle: ContentSearchBundle, metadata: Set<Attribute>): List<Long>

    fun searchBookIdsUniversal(searchBundle: ContentSearchBundle): List<Long>


    fun selectRecentBooks(searchBundle: ContentSearchBundle): LiveData<PagedList<Content>>

    fun searchBooks(
        searchBundle: ContentSearchBundle,
        metadata: Set<Attribute>
    ): LiveData<PagedList<Content>>

    fun searchBooksUniversal(searchBundle: ContentSearchBundle): LiveData<PagedList<Content>>


    fun selectErrorContentLive(): LiveData<List<Content>>

    fun selectErrorContentLive(query: String?, source: Site?): LiveData<List<Content>>

    fun selectErrorContent(): List<Content>


    fun countBooks(
        groupId: Long,
        metadata: Set<Attribute>?,
        location: Location,
        contentType: Type
    ): LiveData<Int>

    fun countAllBooksLive(): LiveData<Int>


    // IMAGEFILES
    fun insertImageFile(img: ImageFile)

    fun insertImageFiles(imgs: List<ImageFile>)

    fun replaceImageList(contentId: Long, newList: List<ImageFile>)

    fun updateImageContentStatus(
        contentId: Long,
        updateFrom: StatusContent?,
        updateTo: StatusContent
    )

    fun updateImageFileStatusParamsMimeTypeUriSize(image: ImageFile)

    fun deleteImageFiles(imgs: List<ImageFile>)

    fun selectImageFile(id: Long): ImageFile?

    fun selectImageFiles(ids: LongArray): List<ImageFile>

    fun selectChapterImageFiles(ids: LongArray): List<ImageFile>

    fun flagImagesForDeletion(ids : LongArray, value : Boolean)

    fun selectDownloadedImagesFromContentLive(id: Long): LiveData<List<ImageFile>>

    fun selectDownloadedImagesFromContent(id: Long): List<ImageFile>

    fun countProcessedImagesById(contentId: Long): Map<StatusContent, Pair<Int, Long>>

    fun selectAllFavouritePagesLive(): LiveData<List<ImageFile>>

    fun countAllFavouritePagesLive(): LiveData<Int>

    fun selectPrimaryMemoryUsagePerSource(): Map<Site, Pair<Int, Long>>

    fun selectPrimaryMemoryUsagePerSource(rootPath: String): Map<Site, Pair<Int, Long>>

    fun selectExternalMemoryUsagePerSource(): Map<Site, Pair<Int, Long>>


    // QUEUE
    fun selectQueue(): List<QueueRecord>

    fun selectQueueLive(): LiveData<List<QueueRecord>>

    fun selectQueueLive(query: String?, source: Site?): LiveData<List<QueueRecord>>

    fun selectQueueUrls(site: Site): Set<String>

    fun addContentToQueue(
        content: Content,
        sourceImageStatus: StatusContent?,
        targetImageStatus: StatusContent?,
        position: QueuePosition,
        replacedContentId: Long,
        replacementTitle: String?,
        isQueueActive: Boolean
    )

    fun updateQueue(queue: List<QueueRecord>)

    fun deleteQueueRecordsCore()

    fun deleteQueue(content: Content)

    fun deleteQueue(index: Int)


    // ATTRIBUTES
    fun insertAttribute(attr: Attribute): Long

    fun selectAttribute(id: Long): Attribute?

    fun selectAttributeMasterDataPaged(
        types: List<AttributeType>,
        filter: String?,
        groupId: Long,
        attrs: Set<Attribute>?,
        location: Location,
        contentType: Type,
        includeFreeAttrs: Boolean,
        page: Int,
        booksPerPage: Int,
        orderStyle: Int
    ): AttributeQueryResult

    fun countAttributesPerType(
        groupId: Long,
        filter: Set<Attribute>?,
        location: Location,
        contentType: Type
    ): SparseIntArray


    // CHAPTERS
    fun selectChapters(contentId: Long): List<Chapter>

    fun selectChapters(chapterIds: List<Long>): List<Chapter>

    fun selectChapter(chapterId: Long): Chapter?

    fun insertChapters(chapters: List<Chapter>)

    fun deleteChapters(content: Content)

    fun deleteChapter(chapter: Chapter)


    // SITE HISTORY
    fun selectHistory(s: Site): SiteHistory

    fun insertSiteHistory(site: Site, url: String)


    // BOOKMARKS
    fun countAllBookmarks(): Long

    fun selectAllBookmarks(): List<SiteBookmark>

    fun selectBookmarks(s: Site): List<SiteBookmark>

    fun selectHomepage(s: Site): SiteBookmark?

    fun insertBookmark(bookmark: SiteBookmark): Long

    fun insertBookmarks(bookmarks: List<SiteBookmark>)

    fun deleteBookmark(bookmarkId: Long)

    fun deleteAllBookmarks()


    // SEARCH HISTORY
    fun selectSearchRecordsLive(): LiveData<List<SearchRecord>>

    fun insertSearchRecord(record: SearchRecord, limit: Int)

    fun deleteAllSearchRecords()


    // RENAMING RULES
    fun selectRenamingRule(id: Long): RenamingRule?

    fun selectRenamingRulesLive(
        type: AttributeType,
        nameFilter: String?
    ): LiveData<List<RenamingRule>>

    fun selectRenamingRules(type: AttributeType, nameFilter: String?): List<RenamingRule>

    fun insertRenamingRule(rule: RenamingRule): Long

    fun insertRenamingRules(rules: List<RenamingRule>)

    fun deleteRenamingRules(ids: List<Long>)


    // RESOURCES
    fun cleanup()

    fun cleanupOrphanAttributes()

    fun getDbSizeBytes(): Long


    // ONE-TIME USE QUERIES (MIGRATION & CLEANUP)
    fun selectContentIdsWithUpdatableJson(): LongArray
}